### PR TITLE
Adjustments in nRF drivers and ext/hal/nordic to provide support for nRF9160

### PR DESCRIPTION
**Pull request by anangl**
**drivers: i2c: nrfx: Add support for TWIM2 and TWIM3**

Extend the i2c_nrfx_twim driver to support all TWIM instances
available on nRF9160.

**drivers: serial: nrfx: Add support for UARTE2 and UARTE3**

Extend the uart_nrfx_uarte driver to support all UARTE instances
available on nRF9160.

**dts: nrf: Add bindings for CLOCK, POWER, RTC, and TIMER peripherals**

Add dts bindings for nRF CLOCK, POWER, RTC, and TIMER peripherals.

**ext: hal: nordic: Add nrfx_config for nRF9160 SoC**

Add the header file with the nrfx drivers configuration suitable for
the nRF9160 SoC. This file contains also definitions of CMSIS symbols
(like NRF_CLOCK or NRF_SPIM0) for structures of peripheral registers.
These symbols are used in nrfx HALs and drivers, and for other nRF SoCs
they come from MDK files. For nRF9160 they need to be mapped to proper
addresses in Secure or Non-Secure domain, so we use the information
coming from DT to get it done automatically.

**dts: nrf9160: Add rtc0 node and replace nvmc node with alias**

Add missing node for instance 0 of the RTC peripheral and use an alias
to the flash-controller node instead of the nvmc node definition (there
is no need to have a separate node just to get the NVMC base address).

**drivers: nrf: Adjust clock_control and timer drivers for nRF9160**

Minor adjustments are done to the nRF clock_control and rtc_timer
drivers to make them usable on nRF9160 as well.

### CODE REVIEW

**Comment by anangl** [comment]
Removed one stray space.

File: dts/arm/nordic/nrf9160_common.dtsi [hunk 1]
```diff
@@ -161,6 +154,13 @@ gpio0: gpio@842500 {
 	status = "disabled";
 };
 
+rtc0: rtc@14000 {
```
Comment (ioannisg) [inline dts/arm/nordic/nrf9160_common.dtsi]: Why was this needed?

Comment (ioannisg) [reply dts/arm/nordic/nrf9160_common.dtsi]: Also, no alias is defined for this one....

Comment (anangl) [reply dts/arm/nordic/nrf9160_common.dtsi]: > Why was this needed?

To get the base address of this peripheral in [nrfx_config_nrf9160.h](https://github.com/zephyrproject-rtos/zephyr/pull/12531/files#diff-7c92595657f7db412148884cfa771899R80).

> Also, no alias is defined for this one....

Right. This is missing. I'll add the alias.

File: drivers/clock_control/nrf_power_clock.c [hunk 2]
```diff
@@ -11,12 +11,16 @@
...
 #include <misc/__assert.h>
-#include "nrf_clock.h"
+#include <nrf_clock.h>
 #if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
 #include <nrf_power.h>
 #include <drivers/clock_control/nrf_clock_control.h>
 #endif
 
+#if defined(CONFIG_SOC_SERIES_NRF91X)
+#define POWER_CLOCK_IRQn  CLOCK_POWER_IRQn
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply drivers/clock_control/nrf_power_clock.c]: Great. Do you think this should be done in this PR?

File: drivers/clock_control/nrf_power_clock.c [hunk 3]
```diff
@@ -254,6 +259,7 @@ static int _k32src_start(struct device *dev, clock_control_subsys_t sub_system)
 			__ASSERT_NO_MSG(err == -EINPROGRESS);
 		}
 	}
+#endif /* NRF_CLOCK_HAS_CALIBRATION */
```
[the remark this answers is not in the document]
Comment (anangl) [reply drivers/clock_control/nrf_power_clock.c]: Because when the corresponding `#if` is only a couple of lines above and it can be easily identified without the help of such comment, there is no need to add one.
In places like this:
```
#if NRF_CLOCK_HAS_CALIBRATION
	u8_t ctto, done;
	struct device *dev = arg;
#endif
#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
	bool usb_detected, usb_pwr_rdy, usb_removed;
#endif
```
adding of such comments after `#endif` does not necessarily improve the code readability:
```
#if NRF_CLOCK_HAS_CALIBRATION
	u8_t ctto, done;
	struct device *dev = arg;
#endif /* NRF_CLOCK_HAS_CALIBRATION */
#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
	bool usb_detected, usb_pwr_rdy, usb_removed;
#endif /* defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840) */
```

File: drivers/clock_control/nrf_power_clock.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (anangl) [reply drivers/clock_control/nrf_power_clock.c]: I can do it here. In a separate commit, to make it more visible.

**Comment by anangl** [comment]
I am not sure how should I fix the CI [issue](https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/31284/1/console).
BT simulator tests cannot be built because of undeclared symbol `NRF_CLOCK_HFCLK_HIGH_ACCURACY`. The problem is that the symbol is declared [in _<nrf_clock.h>_](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/hal/nrf_clock.h#L119) but the simulator uses its own, [stripped-down version of this file](https://github.com/BabbleSim/ext_NRF52_hw_models/blob/master/src/nrfx/hal/nrf_clock.h) and hence does not have it. @aescolar Could you advise?

File: drivers/timer/nrf_rtc_timer.c [hunk 4]
```diff
@@ -49,7 +49,7 @@ static u32_t counter(void)
  * it by pointer at runtime, maybe?) so we don't have this leaky
  * symbol.
  */
-void rtc1_nrf5_isr(void *arg)
+void rtc1_nrf_isr(void *arg)
```
Comment (ioannisg) [inline drivers/timer/nrf_rtc_timer.c]: could this function be declared as static?

Comment (anangl) [reply drivers/timer/nrf_rtc_timer.c]: I guess not. Because the arm_irq_vector_table [apparently uses it](https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/kernel/arm_irq_vector_table/src/arm_irq_vector_table.c#L123). I need to rename it in this test as well.

File: dts/arm/nordic/nrf9160_common.dtsi (hunk 1 above)
Comment (anangl) [reply dts/arm/nordic/nrf9160_common.dtsi]: Done.

File: ext/hal/nordic/nrfx_config_nrf9160.h [hunk 5]
```diff
@@ -0,0 +1,2147 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NRFX_CONFIG_NRF9160_H__
+#define NRFX_CONFIG_NRF9160_H__
+
+// The definitions below
```
[the remark this answers is not in the document]
Comment (anangl) [reply ext/hal/nordic/nrfx_config_nrf9160.h]: Done.

File: drivers/clock_control/nrf_power_clock.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (anangl) [reply drivers/clock_control/nrf_power_clock.c]: Done.

File: drivers/timer/nrf_rtc_timer.c (hunk 4 above)
Comment (anangl) [reply drivers/timer/nrf_rtc_timer.c]: Done (renaming of the function in the test).

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
Belongs to partent issue #12199

**Review by carlescufi — APPROVED** [review]
LGTM

**Comment by ioannisg** [comment]
Somehow the CI needs to pass, though

**Review by lemrey — APPROVED** [review]
LGTM, have done a quick check locally.

**Comment by ioannisg** [comment]
@anangl I believe some of the failures are due to the use of DT_NORDIC_NRF_CLOCK_0_IRQ_0, throughout the clock-control driver. There is no DTS info for clock-control yet for nrf51,52, so these defines are not auto-generated, as far as I understand

**Comment by aescolar** [comment]
> BT simulator tests cannot be built because of undeclared symbol NRF_CLOCK_HFCLK_HIGH_ACCURACY. The problem is that the symbol is declared in <nrf_clock.h> but the simulator uses its own, stripped-down version of this file and hence does not have it. @aescolar Could you advise?

@anangl : (The nRF52 HW models use their own HAL version) Should be easy to fix, I should be able to push a patch tomorrow to align it with the latest nRFx HAL.
Note that there is also a DT define issue (both in the nrf52_bsim and in a few real boards)

**Comment by anangl** [comment]
> Note that there is also a DT define issue (both in the nrf52_bsim and in a few real boards)

Right, my bad. Fixed.

**Comment by anangl** [comment]
@ioannisg
> @anangl I believe some of the failures are due to the use of DT_NORDIC_NRF_CLOCK_0_IRQ_0, throughout the clock-control driver. There is no DTS info for clock-control yet for nrf51,52, so these defines are not auto-generated, as far as I understand

You're right. I fixed that by adding clock nodes for nRF5 family SoCs. Please, take a look.

**Comment by anangl** [comment]
> @ioannisg
> 
> > @anangl I believe some of the failures are due to the use of DT_NORDIC_NRF_CLOCK_0_IRQ_0, throughout the clock-control driver. There is no DTS info for clock-control yet for nrf51,52, so these defines are not auto-generated, as far as I understand
> 
> You're right. I fixed that by adding clock nodes for nRF5 family SoCs. Please, [take a look](https://github.com/zephyrproject-rtos/zephyr/compare/f2f2961bac3541d056b53f3aa44c325019cf1d6a..f219a118b6e90d3990bd9590fdfc24196111378c).

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
recheck

**Comment by anangl** [comment]
> recheck

It will still fail. We must wait until @aescolar updates _nrf_clock.h_ in [ext_NRF52_hw_models](https://github.com/BabbleSim/ext_NRF52_hw_models/blob/master/src/nrfx/hal/nrf_clock.h).

**Comment by anangl** [comment]
Added adjustments needed to utilize PWM peripherals.

**Comment by ioannisg** [comment]
> 
> 
> > BT simulator tests cannot be built because of undeclared symbol NRF_CLOCK_HFCLK_HIGH_ACCURACY. The problem is that the symbol is declared in <nrf_clock.h> but the simulator uses its own, stripped-down version of this file and hence does not have it. @aescolar Could you advise?
> 
> @anangl : (The nRF52 HW models use their own HAL version) Should be easy to fix, I should be able to push a patch tomorrow to align it with the latest nRFx HAL.
> Note that there is also a DT define issue (both in the nrf52_bsim and in a few real boards)

@aescolar the DT issue is (hopefully) addressed.

File: dts/arm/nordic/nrf9160_common.dtsi [hunk 6]
```diff
@@ -152,6 +145,38 @@ spi3: spi@b000 {
 	label = "SPI_3";
 };
 
+pwm0: pwm@21000 {
```
Comment (ioannisg) [inline dts/arm/nordic/nrf9160_common.dtsi]: This has only be added for nRF9160, right?

Comment (anangl) [reply dts/arm/nordic/nrf9160_common.dtsi]: I am not sure if I correctly understood the question.
These nodes were missing for nRF9160. nRF5 family has them already defined (and all the other stuff, like bindings, is already in place).

Comment (ioannisg) [reply dts/arm/nordic/nrf9160_common.dtsi]: Ah right - sorry, I did not check thoroughly.

**Comment by anangl** [comment]
Forgot to add one PWM related commit in the last update.

**Review by aescolar — COMMENTED** [review]
If you are changing the POWER_CLOCK_IRQn , seems you forgot nrf_power_clock.c : 484

**Comment by anangl** [comment]
> If you are changing the POWER_CLOCK_IRQn , seems you forgot nrf_power_clock.c : 484

How could that happen? I was sure I had replaced this one. Anyway, fixed.
Thanks a lot for spotting this!

**Comment by aescolar** [comment]
> If you are changing the POWER_CLOCK_IRQn 

@anangl : For curiosity, what is the motivation to change from using that MDK type into using the DTS macro?

**Comment by anangl** [comment]
> > If you are changing the POWER_CLOCK_IRQn
> 
> @anangl : For curiosity, what is the motivation to change from using that MDK type into using the DTS macro?

@aescolar For nRF9160 this enum value is named CLOCK_POWER_IRQn, and we wanted to avoid using renaming like [this](
 https://github.com/zephyrproject-rtos/zephyr/compare/eda56e292d9d50d031fdd01fb20f123fb3f25d5c..f2f2961bac3541d056b53f3aa44c325019cf1d6a#diff-3c5c76aad21e56267c253f5caa8c09e8L20) (this was the initially proposed change in this PR). Another alternative could be `NRFX_IRQ_NUMBER_GET(NRF_CLOCK)`. I thought the solution with DT symbol would be better.

**Comment by aescolar** [comment]
> For nRF9160 this enum value is named CLOCK_POWER_IRQn

Oh well, I guess the issue then was not being coherent in the MDK side..

**Comment by aescolar** [comment]
recheck

**Comment by ioannisg** [comment]
This one is good to go, IMO; nrfx_config_nrf9160.h has the standard ext/hal/nordic license.

**Review by carlescufi — APPROVED** [review]
